File: Web/src/store/subscriptionStore.ts
import { create } from "zustand";
import { persist } from "zustand/middleware";
import { supabase } from "@/integrations/supabase/client";

export interface SubscriptionPlan {
  id: string;
  name: string;
  price: number;
  monthlyLimit: number;
  stripePriceId?: string;
}

export interface SubscriptionState {
  currentPlan: SubscriptionPlan;
  usageCount: number;
  subscriptionStatus: "free" | "active" | "cancelled" | "past_due";
  subscriptionEnd?: string;
  isLoading: boolean;

  // Actions
  setCurrentPlan: (plan: SubscriptionPlan) => void;
  incrementUsage: () => void;
  resetUsage: () => void;
  setSubscriptionStatus: (
    status: "free" | "active" | "cancelled" | "past_due"
  ) => void;
  setSubscriptionEnd: (date?: string) => void;
  canGenerateImage: () => boolean;
  getRemainingGenerations: () => number;
  checkSubscription: (clerkToken?: string) => Promise<void>;
  createCheckoutSession: (
    priceId: string,
    planName: string,
    clerkToken: string
  ) => Promise<string>;
  openCustomerPortal: (clerkToken: string) => Promise<string>;
  setLoading: (loading: boolean) => void;
  syncWithDatabase: (userId: string) => Promise<void>;
}

export const SUBSCRIPTION_PLANS: SubscriptionPlan[] = [
  {
    id: "free",
    name: "Free",
    price: 0,
    monthlyLimit: 5,
  },
  {
    id: "pro",
    name: "Pro",
    price: 10,
    monthlyLimit: 25,
    stripePriceId: "price_1Rdi4uFA8pQOwelxe6JHX3a5",
  },
  {
    id: "pro_plus",
    name: "Pro Plus",
    price: 50,
    monthlyLimit: 500,
    stripePriceId: "price_1Rdi7HFA8pQOwelxB2on1D0R",
  },
];

export const useSubscriptionStore = create<SubscriptionState>()(
  persist(
    (set, get) => ({
      currentPlan: SUBSCRIPTION_PLANS[0], // Start with free plan
      usageCount: 0,
      subscriptionStatus: "free",
      subscriptionEnd: undefined,
      isLoading: false,

      setCurrentPlan: (plan) => set({ currentPlan: plan }),

      incrementUsage: async () => {
        const state = get();
        const newUsage = state.usageCount + 1;
        set({ usageCount: newUsage });

        // Update usage in database
        try {
          const token = localStorage.getItem("clerk-db-jwt");
          if (token) {
            console.log(JSON.parse(atob(token.split(".")[1])).sub);
            await supabase
              .from("user_subscription")
              .update({
                current_usage: newUsage,
                updated_at: new Date().toISOString(),
              })
              .eq("clerk_user_id", JSON.parse(atob(token.split(".")[1])).sub);
          }
          // Get the current user's email from Clerk
          const userEmail = await getCurrentUserEmail();
          if (!userEmail) {
            console.error("No user email found for incrementUsage");
            return;
          }

          console.log("Syncing subscription for email:", userEmail);
          await supabase
            .from("user_subscription")
            .update({
              current_usage: newUsage,
              updated_at: new Date().toISOString(),
            })
            .eq("email", userEmail);
        } catch (error) {
          console.error("Failed to update usage in database:", error);
        }
      },

      resetUsage: () => set({ usageCount: 0 }),

      setSubscriptionStatus: (status) => set({ subscriptionStatus: status }),

      setSubscriptionEnd: (date) => set({ subscriptionEnd: date }),

      setLoading: (loading) => set({ isLoading: loading }),

      canGenerateImage: () => {
        const state = get();
        return state.usageCount < state.currentPlan.monthlyLimit;
      },

      getRemainingGenerations: () => {
        const state = get();
        return Math.max(0, state.currentPlan.monthlyLimit - state.usageCount);
      },

      syncWithDatabase: async (userId: string) => {
        try {
          // Get the current user's email from Clerk
          const userEmail = await getCurrentUserEmail();
          if (!userEmail) {
            console.error("No user email found for syncing subscription");
            return;
          }

          console.log("Syncing subscription for email:", userEmail);

          // Query subscribers table using email
          const { data, error } = await supabase
            //.from('subscribers')
            .from("user_subscription")
            .select("*")
            .eq("email", userEmail)
            .single();

          if (error && error.code !== "PGRST116") {
            console.error(
              "Error fetching subscription from subscribers table:",
              error
            );
            return;
          }

          if (data) {
            console.log("Found subscription data:", data);

            // Map subscription tier to plan
            const plan =
              SUBSCRIPTION_PLANS.find(
                (p) => p.name === data.subscription_tier
              ) || SUBSCRIPTION_PLANS[0];

            // Determine subscription status
            let status: "free" | "active" | "cancelled" | "past_due" = "free";
            if (data.subscription_status === "active") {
              status = "active";
            } else if (
              data.subscription_tier &&
              data.subscription_tier !== "Free"
            ) {
              status = "cancelled";
            }

            set({
              currentPlan: plan,
              subscriptionStatus: status,
              subscriptionEnd: data.subscription_end || undefined,
            });

            // Also sync with user_subscription table to get usage data
            const { data: userSubData } = await supabase
              .from("user_subscription")
              .select("current_usage")
              .eq("clerk_user_id", userId)
              .single();

            if (userSubData) {
              set({ usageCount: userSubData.current_usage });
            }

            console.log("Subscription synced successfully:", {
              plan: plan.name,
              status,
              subscriptionEnd: data.subscription_end,
            });
          } else {
            console.log("No subscription found, setting to free plan");
            set({
              currentPlan: SUBSCRIPTION_PLANS[0], // Free plan
              subscriptionStatus: "free",
              subscriptionEnd: undefined,
            });
          }
        } catch (error) {
          console.error("Error syncing with database:", error);
        }
      },

      checkSubscription: async (clerkToken?: string) => {
        try {
          set({ isLoading: true });

          if (!clerkToken) {
            console.log("No Clerk token provided");
            return;
          }

          const { data, error } = await supabase.functions.invoke(
            "check-subscription",
            {
              headers: {
                Authorization: `Bearer ${clerkToken}`,
              },
            }
          );

          if (error) {
            console.error("Error checking subscription:", error);
            return;
          }

          if (data.subscribed && data.subscription_tier) {
            const plan = SUBSCRIPTION_PLANS.find(
              (p) => p.name === data.subscription_tier
            );
            if (plan) {
              set({
                currentPlan: plan,
                subscriptionStatus: "active",
                subscriptionEnd: data.subscription_end,
              });
              console.log("checkSubscription:", {
                plan: plan.name,
                status,
                subscriptionEnd: data.subscription_end,
              });
            }
          } else {
            set({
              currentPlan: SUBSCRIPTION_PLANS[0], // Free plan
              subscriptionStatus: "free",
              subscriptionEnd: undefined,
            });
          }
        } catch (error) {
          console.error("Error in checkSubscription:", error);
        } finally {
          set({ isLoading: false });
        }
      },

      createCheckoutSession: async (
        priceId: string,
        planName: string,
        clerkToken: string
      ) => {
        const { data, error } = await supabase.functions.invoke(
          "create-checkout",
          {
            body: { priceId, planName },
            headers: {
              Authorization: `Bearer ${clerkToken}`,
            },
          }
        );

        if (error) {
          throw new Error(error.message || "Failed to create checkout session");
        }

        return data.url;
      },

      openCustomerPortal: async (clerkToken: string) => {
        const { data, error } = await supabase.functions.invoke(
          "customer-portal",
          {
            headers: {
              Authorization: `Bearer ${clerkToken}`,
            },
          }
        );

        if (error) {
          throw new Error(error.message || "Failed to open customer portal");
        }

        return data.url;
      },
    }),
    {
      name: "subscription-storage",
    }
  )
);

// Helper function to get current user email from Clerk
async function getCurrentUserEmail(): Promise<string | null> {
  try {
    // Try to get email from window.Clerk
    if (typeof window !== "undefined" && (window as any).Clerk?.user) {
      const user = (window as any).Clerk.user;
      return (
        user.primaryEmailAddress?.emailAddress ||
        user.emailAddresses?.[0]?.emailAddress ||
        null
      );
    }

    // Fallback: try to get from localStorage or other Clerk storage
    if (typeof window !== "undefined") {
      const clerkSession = localStorage.getItem("clerk-db-jwt");
      if (clerkSession) {
        try {
          const decoded = JSON.parse(atob(clerkSession.split(".")[1]));
          return decoded.email || null;
        } catch (e) {
          console.warn("Could not decode Clerk session for email");
        }
      }
    }

    return null;
  } catch (error) {
    console.error("Error getting current user email:", error);
    return null;
  }
}

File: iOS/Photox/SubscriptionStore.swift
import Foundation
import Supabase
import Clerk

@MainActor
class SubscriptionStore: ObservableObject {
    @Published var currentPlan: SubscriptionPlan = SUBSCRIPTION_PLANS[0]
    @Published var usageCount: Int = 0
    @Published var subscriptionStatus: SubscriptionStatus = .free
    @Published var subscriptionEnd: Date?
    @Published var isLoading: Bool = false

    enum SubscriptionStatus: String {
        case free, active, cancelled, past_due
    }

    func setCurrentPlan(plan: SubscriptionPlan) {
        self.currentPlan = plan
    }

    func incrementUsage() async {
        usageCount += 1
        // Update usage in the database
        do {
            guard let userId = Clerk.shared.user?.id else {
                print("User not authenticated")
                return
            }
            
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            
            try await supabase.database
                .from("user_subscription")
                .update(["current_usage": "\(usageCount)", "updated_at": formatter.string(from: Date())])
                .eq("clerk_user_id", value: userId)
                .execute()
        } catch {
            print("Failed to update usage in database: \(error)")
        }
    }

    func resetUsage() {
        usageCount = 0
    }

    func setSubscriptionStatus(status: SubscriptionStatus) {
        self.subscriptionStatus = status
    }

    func setSubscriptionEnd(date: Date?) {
        self.subscriptionEnd = date
    }

    func canGenerateImage() -> Bool {
        return usageCount < currentPlan.monthlyLimit
    }

    func getRemainingGenerations() -> Int {
        return max(0, currentPlan.monthlyLimit - usageCount)
    }

    func syncWithDatabase() async {
        guard let userId = Clerk.shared.user?.id, let userEmail = Clerk.shared.user?.primaryEmailAddress else {
            print("User not authenticated for syncing subscription")
            return
        }

        print("Syncing subscription for email: \(userEmail)")

        do {
            let userSubscription: UserSubscription = try await supabase.database
                .from("user_subscription")
                .select()
                .eq("email", value: userEmail.emailAddress)
                .single()
                .execute()
                .value

            let plan = SUBSCRIPTION_PLANS.first { $0.name == userSubscription.subscription_tier } ?? SUBSCRIPTION_PLANS[0]
            
            var status: SubscriptionStatus = .free
            if userSubscription.subscription_status == "active" {
                status = .active
            } else if userSubscription.subscription_tier != nil && userSubscription.subscription_tier != "Free" {
                status = .cancelled
            }

            self.currentPlan = plan
            self.subscriptionStatus = status
            self.subscriptionEnd = userSubscription.subscription_end
            self.usageCount = userSubscription.current_usage ?? 0
            
            print("Subscription synced successfully: plan: \(plan.name), status: \(status), subscriptionEnd: \(String(describing: self.subscriptionEnd))")

        } catch {
            print("Error syncing with database, setting to free plan: \(error)")
            self.currentPlan = SUBSCRIPTION_PLANS[0]
            self.subscriptionStatus = .free
            self.subscriptionEnd = nil
        }
    }

    func checkSubscription() async {
        self.isLoading = true
        defer { self.isLoading = false }
        
        guard let token = try? await Clerk.shared.session?.getToken() else {
            print("User not authenticated for checking subscription")
            return
        }

        do {
            let response: CheckSubscriptionResponse = try await supabase.functions.invoke(
                "check-subscription",
                 options: .init(headers: ["Authorization": "Bearer \(token)"])
            )
            
            if response.subscribed, let tier = response.subscription_tier, let plan = SUBSCRIPTION_PLANS.first(where: { $0.name == tier }) {
                self.currentPlan = plan
                self.subscriptionStatus = .active
                if let endDateString = response.subscription_end {
                    let formatter = ISO8601DateFormatter()
                    self.subscriptionEnd = formatter.date(from: endDateString)
                }
                print("checkSubscription: plan: \(plan.name), status: active, subscriptionEnd: \(String(describing: self.subscriptionEnd))")
            } else {
                self.currentPlan = SUBSCRIPTION_PLANS[0]
                self.subscriptionStatus = .free
                self.subscriptionEnd = nil
            }
        } catch {
            print("Error in checkSubscription: \(error)")
        }
    }

    func createCheckoutSession(priceId: String, planName: String) async throws -> CreateCheckoutResponse {
        guard let token = try? await Clerk.shared.session?.getToken() else {
            throw URLError(.userAuthenticationRequired)
        }
        
        let body = CreateCheckoutRequest(priceId: priceId, planName: planName)
        let response: CreateCheckoutResponse = try await supabase.functions.invoke(
            "create-checkout",
            options: .init(
                headers: ["Authorization": "Bearer \(token)"],
                body: body
            )
        )
        return response
    }
    
    func createPaymentIntent(priceId: String, planName: String, amount: Double) async throws -> CreatePaymentIntentResponse {
        guard let token = try? await Clerk.shared.session?.getToken() else {
            throw URLError(.userAuthenticationRequired)
        }
        
        let body = CreatePaymentIntentRequest(priceId: priceId, planName: planName, amount: amount)
        let response: CreatePaymentIntentResponse = try await supabase.functions.invoke(
            "create-payment-intent",
            options: .init(
                headers: ["Authorization": "Bearer \(token)"],
                body: body
            )
        )
        return response
    }
    
    func createSubscription(priceId: String, planName: String) async throws -> CreateSubscriptionIntentResponse {
        guard let token = try? await Clerk.shared.session?.getToken() else {
            throw URLError(.userAuthenticationRequired)
        }
        
        let body = CreateSubscriptionIntentRequest(priceId: priceId, planName: planName)
        let response: CreateSubscriptionIntentResponse = try await supabase.functions.invoke(
            "create-subscription",
            options: .init(
                headers: ["Authorization": "Bearer \(token)"],
                body: body
            )
        )
        return response
    }

    func openCustomerPortal() async throws -> String {
        guard let token = try? await Clerk.shared.session?.getToken() else {
            throw URLError(.userAuthenticationRequired)
        }
        
        let response: CustomerPortalResponse = try await supabase.functions.invoke(
            "customer-portal",
            options: .init(headers: ["Authorization": "Bearer \(token)"])
        )
        return response.url
    }
}

struct SubscriptionPlan: Identifiable {
    let id: String
    let name: String
    let price: Double
    let monthlyLimit: Int
    let stripePriceId: String?
}

let SUBSCRIPTION_PLANS: [SubscriptionPlan] = [
    SubscriptionPlan(id: "free", name: "Free", price: 0, monthlyLimit: 5, stripePriceId: nil),
    SubscriptionPlan(id: "pro", name: "Pro", price: 10, monthlyLimit: 25, stripePriceId: "price_1Rdi4uFA8pQOwelxe6JHX3a5"),
    SubscriptionPlan(id: "pro_plus", name: "Pro Plus", price: 50, monthlyLimit: 500, stripePriceId: "price_1Rdi7HFA8pQOwelxB2on1D0R")
]

struct UserSubscription: Decodable {
    var email: String
    var subscription_tier: String?
    var subscription_status: String?
    var subscription_end: Date?
    var current_usage: Int?
}

struct CheckSubscriptionResponse: Decodable {
    let subscribed: Bool
    let subscription_tier: String?
    let subscription_end: String?
}

struct CreateCheckoutRequest: Encodable {
    let priceId: String
    let planName: String
}

struct CreatePaymentIntentRequest: Encodable {
    let priceId: String
    let planName: String
    let amount: Double
}


struct CreateCheckoutResponse: Decodable {
    let paymentIntent: String
    let ephemeralKey: String
    let customer: String
    let publishableKey: String
}

struct CustomerPortalResponse: Decodable {
    let url: String
}

struct CreatePaymentIntentResponse: Decodable {
    let client_secret: String
    let customer: String
    let publishableKey: String
}

struct CreateSubscriptionIntentRequest: Encodable {
    let priceId: String
    let planName: String
}

struct CreateSubscriptionIntentResponse: Decodable {
    let clientSecret: String
    let customer: String
    let publishableKey: String
}
